### PR TITLE
Dump swapchain image to file

### PR DESCRIPTION
**Pull request by RenfengLiu**
Dump the swapchain images to ppm file. This is controlled by
environment variable "IMAGE_DUMP_PATH".

### CODE REVIEW

File: core/vulkan/vk_virtual_swapchain/cc/virtual_swapchain.cpp [hunk 1]
```diff
@@ -294,6 +305,64 @@ void VirtualSwapchain::Destroy(const VkAllocationCallbacks* pAllocator) {
...
+      default:
+        std::cout << "Format " << image_format
+                  << " not supported for dump yet.";
+        break;
+    }
+    out.close();
+  };
+
+  std::thread file_writer(ppm_writer, std::move(image_data_owned), size,
+                          std::move(image_path), width_, height_,
```
Comment (purvisa-at-google-com) [inline core/vulkan/vk_virtual_swapchain/cc/virtual_swapchain.cpp]: Nit: Not sure I'd bother with the std::move on the string!

File: core/vulkan/vk_virtual_swapchain/cc/virtual_swapchain.cpp [hunk 2]
```diff
@@ -294,6 +305,64 @@ void VirtualSwapchain::Destroy(const VkAllocationCallbacks* pAllocator) {
...
+  auto image_path = image_dump_dir_ + "/image_" +
+                    std::to_string(dumped_frame_count_++) + "_ts_" +
+                    std::to_string(now) + ".ppm";
+
+  auto ppm_writer = [](std::unique_ptr<char[]> image_data, size_t size,
+                       std::string file_name, uint32_t width, uint32_t height,
+                       VkFormat image_format) {
+    std::ofstream out(file_name, std::ios::out | std::ios::binary);
+    if (!out) {
+      std::cout << "Create file " << file_name << " failed." << std::endl;
```
Comment (purvisa-at-google-com) [inline core/vulkan/vk_virtual_swapchain/cc/virtual_swapchain.cpp]: cerr?

File: core/vulkan/vk_virtual_swapchain/cc/virtual_swapchain.cpp [hunk 3]
```diff
@@ -294,6 +305,64 @@ void VirtualSwapchain::Destroy(const VkAllocationCallbacks* pAllocator) {
...
+          uint32_t* row = (uint32_t*)data;
+          for (uint32_t x = 0; x < width; x++) {
+            out.write((char*)row, 3);
+            row++;
+          }
+          data += width * 4;
+        }
+        break;
+      default:
+        std::cout << "Format " << image_format
```
Comment (purvisa-at-google-com) [inline core/vulkan/vk_virtual_swapchain/cc/virtual_swapchain.cpp]: cerr?

File: core/vulkan/vk_virtual_swapchain/cc/virtual_swapchain.cpp [hunk 4]
```diff
@@ -294,6 +305,64 @@ void VirtualSwapchain::Destroy(const VkAllocationCallbacks* pAllocator) {
...
+void VirtualSwapchain::DumpImageToFile(uint8_t* image_data, size_t size) {
+  std::unique_ptr<char[]> image_data_owned(new char[size]());
+  memcpy(image_data_owned.get(), image_data, size);
+
+  auto now = std::chrono::system_clock::now().time_since_epoch().count();
+  auto image_path = image_dump_dir_ + "/image_" +
+                    std::to_string(dumped_frame_count_++) + "_ts_" +
+                    std::to_string(now) + ".ppm";
+
+  auto ppm_writer = [](std::unique_ptr<char[]> image_data, size_t size,
```
Comment (purvisa-at-google-com) [inline core/vulkan/vk_virtual_swapchain/cc/virtual_swapchain.cpp]: Nit: It would be nice to see more empty lines breaking this up for readability,

File: core/vulkan/vk_virtual_swapchain/cc/virtual_swapchain.cpp [hunk 5]
```diff
@@ -294,6 +305,64 @@ void VirtualSwapchain::Destroy(const VkAllocationCallbacks* pAllocator) {
   functions_->vkDestroyCommandPool(device_, command_pool_, pAllocator);
 }
 
+void VirtualSwapchain::DumpImageToFile(uint8_t* image_data, size_t size) {
```
Comment (purvisa-at-google-com) [inline core/vulkan/vk_virtual_swapchain/cc/virtual_swapchain.cpp]: This doesn't look like it writes the alpha out. Could we write that to a second file beside the colour data, if it exists?

Comment (hevrard) [reply core/vulkan/vk_virtual_swapchain/cc/virtual_swapchain.cpp]: Yeah PPM does not support alpha. In Amber we used lodepng (https://lodev.org/lodepng/) to produce PNG, which support alpha. Could this be a better option? (Lodepng might be problematic if performance is an issue here)

File: core/vulkan/vk_virtual_swapchain/cc/virtual_swapchain.cpp (hunk 4 above)
Comment (hevrard) [reply core/vulkan/vk_virtual_swapchain/cc/virtual_swapchain.cpp]: I'm not a fan of a big lambda to encapsulate the whole PPM encoder, could it be possible to have a separate function for it?

Or just use a wrapper around lodepng, if we go that way.

File: core/vulkan/vk_virtual_swapchain/cc/virtual_swapchain.cpp (hunk 2 above)
Comment (hevrard) [reply core/vulkan/vk_virtual_swapchain/cc/virtual_swapchain.cpp]: More generally, std::cout is irrelevant on Android... 
There does not seem to be a single print/log in the virtual swapchain code, maybe for the sake of portability, so maybe it's OK to not output anything here?

File: core/vulkan/vk_virtual_swapchain/cc/virtual_swapchain.cpp (hunk 5 above)
Comment (purvisa-at-google-com) [reply core/vulkan/vk_virtual_swapchain/cc/virtual_swapchain.cpp]: Would recommend as a follow up commit, not an extension to this one.

File: core/vulkan/vk_virtual_swapchain/cc/BUILD.bazel [hunk 6]
```diff
@@ -46,6 +46,7 @@ cc_library(
     visibility = ["//visibility:public"],
     deps = [
         "//core/vulkan/cc/include/vulkan",
+        "@loadpng",
```
Comment (hevrard) [inline core/vulkan/vk_virtual_swapchain/cc/BUILD.bazel]: Should this be `@lodepng` ?

File: tools/build/workspace.bzl [hunk 7]
```diff
@@ -58,7 +58,7 @@ def gapid_dependencies(android = True, mingw = True, locals = {}):
         locals = locals,
         organization = "google",
         project = "protobuf",
-        commit = "815ff7e1fb2d417d5aebcbf5fc46e626b18dc834", # Head of 3.8.x branch
+        commit = "815ff7e1fb2d417d5aebcbf5fc46e626b18dc834",  # Head of 3.8.x branch
```
Comment (hevrard) [inline tools/build/workspace.bzl]: Extra space?

(this may be due to some auto-formatter of .bzl files)

File: tools/build/workspace.bzl [hunk 8]
```diff
@@ -140,6 +140,15 @@ def gapid_dependencies(android = True, mingw = True, locals = {}):
         sha256 = "3ef3d905849d547b6481b16d8e7b473a84efafbe90131e7bc90a0c6aae4cd8e6",
     )
 
+    maybe_repository(
+        new_git_repository,
+        name = "loadpng",
```
Comment (hevrard) [inline tools/build/workspace.bzl]: I think the name should be `lodepng` to match the repo name

File: tools/build/workspace.bzl [hunk 9]
```diff
@@ -140,6 +140,15 @@ def gapid_dependencies(android = True, mingw = True, locals = {}):
...
+        locals = locals,
+        remote = "https://github.com/lvandeve/lodepng",
+        commit = "3606db5a71359a4165308e7cd5afe67bd42f7edf",
+        build_file = "@gapid//tools/build/third_party:loadpng.BUILD",
```
Comment (hevrard) [inline tools/build/workspace.bzl]: loadpng -> lodepng

File: tools/build/workspace.bzl [hunk 10]
```diff
@@ -245,7 +254,7 @@ def gapid_dependencies(android = True, mingw = True, locals = {}):
         locals = locals,
         url = "https://registry.npmjs.org/vscode-jsonrpc/-/vscode-jsonrpc-2.4.0.tgz",
         build_file = "@gapid//tools/build/third_party:vscode-jsonrpc.BUILD",
-        sha256= "bed9b2facb7d179f14c8a710db8e613be56bd88b2a75443143778813048b5c89",
+        sha256 = "bed9b2facb7d179f14c8a710db8e613be56bd88b2a75443143778813048b5c89",
```
Comment (hevrard) [inline tools/build/workspace.bzl]: fix formatting: let's do it in another PR.

File: tools/build/workspace.bzl (hunk 7 above)
Comment (RenfengLiu) [reply tools/build/workspace.bzl]: yes, they are done automatically by the buildifier

Comment (hevrard) [reply tools/build/workspace.bzl]: Is it possible to isolated them in a separate PR?

I'm pushy on this to avoid later `git log` / `git blame` to have these formatting changes linked to a commit that is about dumping an image in the virtual swapchain.
